x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/support/CacheInvalidatorRegistryTests.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.health.ClusterHealthStatus;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.xpack.security.support.CacheInvalidatorRegistry.CacheInvalidator;
+import org.junit.Before;
+
+import java.time.Instant;
+import java.util.List;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+
+public class CacheInvalidatorRegistryTests extends ESTestCase {
+
+    private CacheInvalidatorRegistry cacheInvalidatorRegistry;
+
+    @Before
+    public void setup() {
+        cacheInvalidatorRegistry = new CacheInvalidatorRegistry();
+    }
+
+    public void testRegistryWillNotAllowInvalidatorsWithDuplicatedName() {
+        cacheInvalidatorRegistry.registerCacheInvalidator("service1", mock(CacheInvalidator.class));
+        final IllegalArgumentException e = expectThrows(
+            IllegalArgumentException.class,
+            () -> cacheInvalidatorRegistry.registerCacheInvalidator("service1", mock(CacheInvalidator.class)));
+        assertThat(e.getMessage(), containsString("already has an entry with name: [service1]"));
+    }
+
+    public void testSecurityIndexStateChangeWillInvalidateAllRegisteredInvalidators() {
+        final CacheInvalidator invalidator1 = mock(CacheInvalidator.class);
+        cacheInvalidatorRegistry.registerCacheInvalidator("service1", invalidator1);
+        final CacheInvalidator invalidator2 = mock(CacheInvalidator.class);
+        cacheInvalidatorRegistry.registerCacheInvalidator("service2", invalidator2);
+
+        final SecurityIndexManager.State previousState = SecurityIndexManager.State.UNRECOVERED_STATE;
+        final SecurityIndexManager.State currentState = new SecurityIndexManager.State(
+            Instant.now(), true, true, true, Version.CURRENT,
+            ".security", ClusterHealthStatus.GREEN, IndexMetadata.State.OPEN);
+
+        cacheInvalidatorRegistry.onSecurityIndexStageChange(previousState, currentState);
+        verify(invalidator1).invalidateAll();
+        verify(invalidator2).invalidateAll();
+    }
+
+    public void testInvalidateByKey() {
```

Comment:
A Nit, with a real implication.

This test name just refers to the method that it tests. It doesn't actually say what behaviour it's testing, just the method it calls. A name like
```suggestion
    public void testInvalidateByKeyCallsCorrectInvalidatorObject() {
```
would be more descriptive.

_And_ the implication is that because we have a method named `test{SomeMethod}` it leads to thinking that we just need to call some stuff on that method and we're done, and skips the thinking step of _What cases do I need to test?_.  
And in this case, there's no tests for invalidating a cache that hasn't been registered. Should that be a no-op? Or a failure?